### PR TITLE
Precinct Selection Scan Fix + Refactor

### PR DESCRIPTION
**Pull request by adghayes**
Fixes bug on main where no ballots will scan in precinct scanner on all precincts mode: https://votingworks.slack.com/archives/CEL6D3GAD/p1661984695650299.

Standardizes the `PrecinctSelection` type across apps, and refactors in other apps where appropriate. I recommend reviewing by app:
- `libs/types`: some additions and edits to to `PrecinctSelection` types
- `libs/ui`: use type in `ElectionInfoBar`
- `services/scan`: use and store `PrecinctSelection` rather than `PrecinctId`
- `frontends/precinct-scanner`: use `PrecinctSelection` rather than `PrecinctId`
- `frontends/bmd`: use type from `libs/types` rather than the frontend types

### CODE REVIEW

File: libs/types/src/precinct_selection.ts [hunk 1]
```diff
@@ -1,22 +1,65 @@
...
+  };
+}
+
+export const ALL_PRECINCTS_SELECTION: AllPrecinctsSelection = {
+  kind: 'AllPrecincts',
+};
+
+export const ALL_PRECINCTS_NAME = 'All Precincts';
+
+export function getPrecinctSelectionName(
```
Comment (arsalansufi) [inline libs/types/src/precinct_selection.ts:48]: This might actually better fit in `libs/utils` - if we put it there, we can also make us of the `find` util instead of manually recreating that logic. The same could be said of `getSinglePrecinctSelection`

File: services/scan/src/importer.ts [hunk 2]
```diff
@@ -136,6 +137,8 @@ export class Importer {
    */
   configure(electionDefinition: ElectionDefinition): void {
     this.workspace.store.setElection(electionDefinition);
+    // Central scanner only uses all precinct mode, set on every configure
+    this.workspace.store.setPrecinctSelection(ALL_PRECINCTS_SELECTION);
```
Comment (arsalansufi) [inline services/scan/src/importer.ts:141]: Since this added logic is specific to the central scanner, we may want to move it up to `services/scan/src/central_scanner_app.ts` instead of keeping it in this lower, machine-agnostic layer

File: services/scan/src/store.ts [hunk 3]
```diff
@@ -57,7 +58,7 @@ export enum ConfigKey {
   MarkThresholdOverrides = 'markThresholdOverrides',
   // @deprecated
   SkipElectionHashCheck = 'skipElectionHashCheck',
-  CurrentPrecinctId = 'currentPrecinctId',
+  PrecinctSelectionKey = 'precinctSelection',
```
Comment (arsalansufi) [inline services/scan/src/store.ts:61]: For consistency with the other keys in this enum, we may want to shorten this to `PrecinctSelection = 'precinctSelection'`. The `Key` in `ConfigKey` should make it clear that this is a key, e.g. `ConfigKey.PrecinctSelection`

File: services/scan/src/workers/interpret_vx.ts [hunk 4]
```diff
@@ -55,6 +59,7 @@ export async function configure(store: Store): Promise<void> {
   debug('creating a new interpreter');
   interpreter = new Interpreter({
     electionDefinition,
+    precinctSelection: ALL_PRECINCTS_SELECTION,
```
Comment (arsalansufi) [inline services/scan/src/workers/interpret_vx.ts:62]: Is there a reason why we're using `ALL_PRECINCTS_SELECTION` here instead of `store.getPrecinctSelection()`?

File: libs/ui/src/election_info_bar.tsx [hunk 5]
```diff
@@ -48,9 +43,15 @@ export function ElectionInfoBar({
         <NoWrap as="strong">{electionDefinition.election.title}</NoWrap> —{' '}
         <NoWrap>{electionDate}</NoWrap>
         <Text as="div" small>
-          {showPrecinctInfo && (
+          {precinctSelection && (
             <React.Fragment>
-              <NoWrap>{precinct?.name ?? 'All Precincts'}</NoWrap> —{' '}
+              <NoWrap>
+                {getPrecinctSelectionName(
```
Comment (arsalansufi) [inline libs/ui/src/election_info_bar.tsx:49]: A slightly more intuitive name for this helper might be `getSelectedPrecinctName` or `precinctSelectionToReadableString`

File: frontends/bmd/src/pages/admin_screen.tsx [hunk 6]
```diff
@@ -57,12 +59,9 @@ export function AdminScreen({
...
 
     if (precinctId === ALL_PRECINCTS_OPTION_VALUE) {
-      updateAppPrecinct({ kind: PrecinctSelectionKind.AllPrecincts });
+      updateAppPrecinct(ALL_PRECINCTS_SELECTION);
     } else if (precinctId) {
-      updateAppPrecinct({
-        kind: PrecinctSelectionKind.SinglePrecinct,
-        precinctId,
-      });
+      updateAppPrecinct(getSinglePrecinctSelection(precinctId));
```
Comment (arsalansufi) [inline frontends/bmd/src/pages/admin_screen.tsx:64]: I don't feel strongly about this! `constructSinglePrecinctSelection` might be a clearer name than `getSinglePrecinctSelection`, which reads as if we're retrieving something from a store

**Review by arsalansufi — COMMENTED** [review]
Thank you for making this consistent across VxSuite! Just a few thoughts!

File: libs/types/src/precinct_selection.ts (hunk 1 above)
Comment (adghayes) [reply libs/types/src/precinct_selection.ts:48]: Will do!

File: services/scan/src/importer.ts (hunk 2 above)
Comment (adghayes) [reply services/scan/src/importer.ts:141]: I originally had it in the higher layer but it had to be in a few places - both on initial setup and any reconfigures - to work properly.

The main point to make is that, I believe, `importer.ts` is not actually machine-agnostic. It is only used for VxCentralScan. I think @jonahkagan told me this. Will let him or @eventualbuddha confirm.

File: services/scan/src/store.ts (hunk 3 above)
Comment (adghayes) [reply services/scan/src/store.ts:61]: I tried this but there's a variable name conflict in the file if you do that, will take another look to see if there's another way.

File: services/scan/src/workers/interpret_vx.ts (hunk 4 above)
Comment (adghayes) [reply services/scan/src/workers/interpret_vx.ts:62]: This is a really helpful catch @arsalansufi, I think this was a mistake.

File: services/scan/src/importer.ts (hunk 2 above)
Comment (adghayes) [reply services/scan/src/importer.ts:141]: I was confused by this at first and if `importer.ts` is machine specific now, we might rename the file to reflect that.

File: services/scan/src/workers/interpret_vx.ts (hunk 4 above)
Comment (arsalansufi) [reply services/scan/src/workers/interpret_vx.ts]: Ah nice!

File: services/scan/src/importer.ts (hunk 2 above)
Comment (arsalansufi) [reply services/scan/src/importer.ts:141]: Ah interesting - I like the idea of renaming then! Maybe `central_scanner_importer.ts` / `CentralScanImporter`

File: frontends/bmd/src/pages/admin_screen.tsx (hunk 6 above)
Comment (adghayes) [reply frontends/bmd/src/pages/admin_screen.tsx]: Renamed to `singlePrecinctSelectionFor`.

File: services/scan/src/workers/interpret_vx.ts (hunk 4 above)
Comment (adghayes) [reply services/scan/src/workers/interpret_vx.ts]: Fixed.

File: libs/ui/src/election_info_bar.tsx (hunk 5 above)
Comment (adghayes) [reply libs/ui/src/election_info_bar.tsx:49]: The current name makes the most sense to me, as `Selected` feels like it implies a selection was made when that was not always the case and `ReadableString` is less clear than `Name` only because `Name` is how we've been identifying these strings in code to date.

File: services/scan/src/store.ts (hunk 3 above)
Comment (adghayes) [reply services/scan/src/store.ts]: Renamed the key by importing the type as:
```
  PrecinctSelection as PrecinctSelectionType,
```

File: services/scan/src/importer.ts (hunk 2 above)
Comment (adghayes) [reply services/scan/src/importer.ts:141]: https://github.com/votingworks/vxsuite/issues/2532

File: libs/ui/src/election_info_bar.tsx (hunk 5 above)
Comment (arsalansufi) [reply libs/ui/src/election_info_bar.tsx:45]: Ah good point that the precinct isn't actually selected on VxCentralScan - the current name is okay then!

**Review by arsalansufi — APPROVED** [review]
LGTM!

File: services/scan/src/importer.ts (hunk 2 above)
Comment (jonahkagan) [reply services/scan/src/importer.ts:140]: Yep, importer is now only used by the central scanner. Renaming it makes sense!

File: services/scan/src/central_scanner_app.ts [hunk 7]
```diff
@@ -197,21 +197,21 @@ export async function buildCentralScannerApp({
     response.json({ status: 'ok' });
   });
 
-  app.get<NoParams, Scan.GetCurrentPrecinctConfigResponse>(
+  app.get<NoParams, Scan.GetPrecinctSelectionConfigResponse>(
```
Comment (jonahkagan) [inline services/scan/src/central_scanner_app.ts:200]: Does the central scanner use these endpoints at all? As far as I can tell, no. And I don't see why it would need them since it's not configurable by precinct. They can probably be removed (along with the corresponding functions in frontends/bsd/src/api/config.ts)

File: services/scan/src/importer.ts [hunk 8]
```diff
@@ -236,17 +241,18 @@ export class Importer {
       backInterpretation
     );
 
-    debug('currentPrecinctId=%s', currentPrecinctId);
-    if (currentPrecinctId) {
+    debug('precinctSelection=%o', precinctSelection);
+    if (precinctSelection.kind !== 'AllPrecincts') {
```
Comment (jonahkagan) [inline services/scan/src/importer.ts:245]: Isn't this case now impossible?

File: services/scan/src/precinct_scanner_app.ts [hunk 9]
```diff
@@ -248,15 +250,15 @@ export async function buildPrecinctScannerApp(
...
       return;
     }
-    store.setCurrentPrecinctId(bodyParseResult.ok().precinctId);
+    store.setPrecinctSelection(bodyParseResult.ok().precinctSelection);
     await configureInterpreter(interpreter, workspace);
     response.json({ status: 'ok' });
   });
 
-  app.delete<NoParams, Scan.DeleteCurrentPrecinctConfigResponse>(
+  app.delete<NoParams, Scan.DeletePrecinctSelectionConfigResponse>(
```
Comment (jonahkagan) [inline services/scan/src/precinct_scanner_app.ts:258]: Is this endpoint used anymore? I noticed you removed it from the tests

File: frontends/precinct-scanner/src/screens/election_manager_screen.tsx [hunk 10]
```diff
@@ -126,6 +137,12 @@ export function ElectionManagerScreen({
     await unconfigure();
   }
 
+  const initialPrecinctValue = precinctSelection
+    ? precinctSelection.kind === 'AllPrecincts'
+      ? ALL_PRECINCTS_OPTION_VALUE
+      : precinctSelection.precinctId
+    : '';
```
Comment (jonahkagan) [inline frontends/precinct-scanner/src/screens/election_manager_screen.tsx:144]: This is a bit of a confusing name. The `Select` is a controlled component, so its value will always be whatever is passed into the `value` prop. So this variable isn't the "initial" value but rather just the value. I might name it `precinctSelectionValue` or something like that

File: frontends/bmd/src/pages/admin_screen.tsx [hunk 11]
```diff
@@ -103,7 +101,7 @@ export function AdminScreen({
                     Select a precinct for this device…
                   </option>
                   <option value={ALL_PRECINCTS_OPTION_VALUE}>
-                    {AllPrecinctsDisplayName}
+                    All Precincts
```
Comment (jonahkagan) [inline frontends/bmd/src/pages/admin_screen.tsx:104]: Didn't you make a constant for this in the precinct selection utils?

File: services/scan/src/workers/interpret_vx.ts (hunk 4 above)
Comment (jonahkagan) [reply services/scan/src/workers/interpret_vx.ts]: @adghayes were the tests passing with this mistake? If so, that might be a bit concerning

**Review by jonahkagan — APPROVED** [review]
Nice work! The refactoring makes things a lot cleaner. Left a few other ideas for cleaning up more. If possible, it would make it easier to review if you can separate out the refactoring into a separate PR next time (or separate commits).

File: frontends/bmd/src/pages/admin_screen.tsx (hunk 11 above)
Comment (adghayes) [reply frontends/bmd/src/pages/admin_screen.tsx:104]: Yes, I just didn't bother importing it here. Can do so.

File: services/scan/src/central_scanner_app.ts (hunk 7 above)
Comment (adghayes) [reply services/scan/src/central_scanner_app.ts:200]: No, we don't use them, wasn't sure if I should remove them. I will go ahead and do so.

File: services/scan/src/precinct_scanner_app.ts (hunk 9 above)
Comment (adghayes) [reply services/scan/src/precinct_scanner_app.ts:258]: It is being used, yes, I'll have to check where I removed it from the tests.

File: services/scan/src/importer.ts (hunk 8 above)
Comment (adghayes) [reply services/scan/src/importer.ts:245]: Yes, this is a place where we still have precinct-specific logic in the central scanner. Like with the endpoint, we can remove them.

File: services/scan/src/precinct_scanner_app.ts (hunk 9 above)
Comment (jonahkagan) [reply services/scan/src/precinct_scanner_app.ts:258]: I guess a more relevant question would be: is there a use case for clearing the VxScan precinct selection back to an unconfigured state? I can't really see why a user would want to do that personally.

Comment (adghayes) [reply services/scan/src/precinct_scanner_app.ts:258]: When you unconfigure the machine, you want it to return to a no precinct state. The precinct selection makes no sense outside of the context of the election definition, so if the election definition is removed, the precinct selection should be too.

Are you suggesting that we delete this endpoint and then just remove the precinct selection on zeroing out the scanner? That makes more sense to me, now that I'm thinking about it.

File: services/scan/src/precinct_scanner_app.test.ts [hunk 12]
```diff
@@ -120,6 +119,14 @@ function postTemplate(
     .expect(200, { status: 'ok' });
 }
 
+async function setAppPrecinct(app: Application, precinctId?: PrecinctId) {
+  await put(app, '/precinct-scanner/config/precinct', {
+    precinctSelection: precinctId
+      ? singlePrecinctSelectionFor(precinctId)
+      : ALL_PRECINCTS_SELECTION,
+  });
+}
```
Comment (jonahkagan) [inline services/scan/src/precinct_scanner_app.test.ts:128]: I realized we don't have a test that confirms that unconfiguring the scanner using the `DELETE /precinct-scanner/config/election` endpoint also clears the precinct config. Reading the code, I _think_ it works, but might be worth checking/testing if you haven't already, since the frontend doesn't explicitly clear the precinct config on unconfigure.

File: services/scan/src/precinct_scanner_app.ts (hunk 9 above)
Comment (jonahkagan) [reply services/scan/src/precinct_scanner_app.ts:258]: Yeah, and I think that's actually what we already do. Just left another comment about it haha

File: frontends/precinct-scanner/src/screens/election_manager_screen.tsx (hunk 10 above)
Comment (adghayes) [reply frontends/precinct-scanner/src/screens/election_manager_screen.tsx:144]: Renamed.

File: frontends/bmd/src/pages/admin_screen.tsx (hunk 11 above)
Comment (adghayes) [reply frontends/bmd/src/pages/admin_screen.tsx]: Switched.

File: services/scan/src/precinct_scanner_app.test.ts (hunk 12 above)
Comment (adghayes) [reply services/scan/src/precinct_scanner_app.test.ts:128]: I don't think we have a test that covers the` DELETE /precinct-scanner/config/election` endpoint at all. Will do some manual testing.

Comment (adghayes) [reply services/scan/src/precinct_scanner_app.test.ts:128]: We should also of course write a test, but I need to close this out and need to do manual testing anyway.

File: services/scan/src/importer.ts (hunk 8 above)
Comment (adghayes) [reply services/scan/src/importer.ts]: Removed.

File: services/scan/src/central_scanner_app.ts (hunk 7 above)
Comment (adghayes) [reply services/scan/src/central_scanner_app.ts]: Removed.

**Comment by adghayes** [comment]
@jonahkagan I addressed most of your feedback. In manual testing, I found that my changes had broken the configuration process for VxScan. It was because `configureInterpreter` now requires a precinct selection, but was being called without one having been made. I changed the logic of when we configure the interpreter in `precinct_scanner_app.ts`. The central scanner hits the `/doneTemplates` endpoint after all templates have been uploaded which triggers a configure. That's what we were doing in `precinct-scanner`, but now that you have to select a precinct, you can't configure the interpreter until after a precinct selection. All automated testing was passing before I made this change.

To your previous question, automated testing was passing before I fixed the mistake that @arsalansufi pointed out. I have not dug into why that was the case yet. 

I'm not going to merge this until I get a LGTM or another close look on your part. This turned out, in general, to be a much deeper dive than expected.

File: services/scan/src/precinct_scanner_app.ts [hunk 13]
```diff
@@ -91,7 +91,7 @@ export async function buildPrecinctScannerApp(
   const { store } = workspace;
 
   // Load interpreter configuration from the store on startup
-  if (store.getElectionDefinition()) {
+  if (store.getElectionDefinition() && store.getPrecinctSelection()) {
```
Comment (jonahkagan) [inline services/scan/src/precinct_scanner_app.ts:94]: Since there are now a bunch of repeated checks that the precinct selection is configured before configuring the interpreter, we could move them inside the `configureInterpreter` function (along with the check for election definition).

**Review by jonahkagan — APPROVED** [review]

File: services/scan/src/precinct_scanner_app.ts (hunk 13 above)
Comment (adghayes) [reply services/scan/src/precinct_scanner_app.ts:94]: I like that, will do one last refactor. Seems like a potential foot gun later if we don't.

**Comment by adghayes** [comment]
Did some manual testing on VxScan (testing different precinct configurations, and reconfigure) and VxCentralScan (testing scanning, and reconfigure). Looks good. Merging.
